posts/2017-09-14-understanding-and-controlling-aws-lambda-costs.md

```diff
@@ -0,0 +1,75 @@
+
+![Cost summary of various serverless FaaS services in US Dollars, as of September 2017. Free tier GB-sec is per month.](https://cdn-images-1.medium.com/max/800/1*CUxf_2hkTNRqlbm-f9rHcQ.png)
+
+Since GB-sec is a composite unit, it can be unintuitive to reason about. Focusing on AWS Lambda’s current pricing, the following chart shows the cost of executing 100,000 invocations of a Lambda function that executes for a varying amount of time, broken down by three different maximum memory sizes. Note that each invocation may have to start a language runtime and load third-party libraries before getting to user code, all of which add to the function's billable runtime **for each invocation**.
+
+![Current Lambda function pricing](https://cdn-images-1.medium.com/max/800/0*kxZTolFAtiBCWr3C.)
+
+It can be helpful to remember that a GB-sec represents neither a gigabyte nor a second, and is the first composite unit to be used for commercial cloud product pricing in recent memory.
+
+Spreadsheet-inclined users may want to look at the data in this [public, read-only Google Sheet](https://docs.google.com/spreadsheets/d/1xh_rNQwGVLos7Fshq2eM4gbIz546LwvrY1fMcLGqPmw/edit?zx=p7c6vgh0j97k#gid=541900098).
+
+## Observations
+
+* **Costs are multiplicative in function memory size and execution time**. Suppose that a Lambda function uses 512 MB of memory and executes in slightly less than 200 milliseconds. After a code change, the function now needs 400 milliseconds to run (double), and 1024 MB of memory (double). The total compute cost increases **4 times**. If the memory requirement is instead tripled to 1536 MB, the total cost increases **6 times**. If both the memory requirement and execution time are tripled, the total cost increases **9 times** ([spreadsheet](https://docs.google.com/spreadsheets/d/1xh_rNQwGVLos7Fshq2eM4gbIz546LwvrY1fMcLGqPmw/edit?zx=p7c6vgh0j97k#gid=541900098)). The impact of multiplicative costs may not be intuitive in that small changes to *either* function memory size or execution time can cause large changes in the total billable cost.
+
+
+* **Processing delays can be expensive**. Suppose that a 512 MB Lambda function uses executes in slightly less than **200 milliseconds**, with a hard limit of 5 seconds. As part of its processing, the function calls an external service over HTTPS and waits for a response before ending. Suppose that network congestion or an external service degradation adds a spike of **2 seconds** to each network call. For the duration of the latency spike, **the extra 2 seconds of Lambda running time will increase cost *11 times***, from \$0.17 per 100k requests to \$1.83 ([spreadsheet](https://docs.google.com/spreadsheets/d/1xh_rNQwGVLos7Fshq2eM4gbIz546LwvrY1fMcLGqPmw/edit?zx=p7c6vgh0j97k#gid=541900098)).
+
```

Comment:
"Processing delays can be expensive. Suppose that a 512 MB Lambda function ~~uses~~ executes in slightly less ..."